reactive/tls_client.py

```diff
@@ -1,101 +1,95 @@
+                log('Writing server key to {0}'.format(key_path))
+                _write_file(key_path, server_key)
 
 
 @when('certificates.client.cert.available')
-@when('tls-client.certificates.directory')
 def store_client(tls):
     '''Read the client certificate and client key from the relation object
     and copy them to the certificate directory.'''
     client_cert, client_key = tls.get_client_cert()
     if client_cert and client_key:
         layer_options = layer.options('tls-client')
-        # Get the certificate directory from the layer options.
-        directory = layer_options.get('certificates-directory')
-        if not os.path.isdir(directory):
-            os.makedirs(directory)
-        cert_file = os.path.join(directory, 'client.crt')
-        hookenv.log('Writing client certificate to {0}'.format(cert_file))
-        with open(cert_file, 'w') as stream:
-            stream.write(client_cert)
-        key_file = os.path.join(directory, 'client.key')
-        hookenv.log('Writing client key to {0}'.format(key_file))
-        with open(key_file, 'w') as stream:
-            stream.write(client_key)
+        if data_changed('client_certificate', client_cert):
+            cert_path = layer_options.get('client_certificate_path')
+            if cert_path:
+                log('Writing client certificate to {0}'.format(cert_path))
+                _write_file(cert_path, client_cert)
+        if data_changed('client_key', client_key):
+            key_path = layer_options.get('client_key_path')
+            if key_path:
+                log('Writing client key to {0}'.format(key_path))
+                _write_file(key_path, client_key)
 
 
 def install_ca(certificate_authority):
     '''Install a certificiate authority on the system by calling the
     update-ca-certificates command.'''
     if certificate_authority:
         name = hookenv.service_name()
-        ca_file = '/usr/local/share/ca-certificates/{0}.crt'.format(name)
-        hookenv.log('Writing CA to {0}'.format(ca_file))
-        # Write the contents of certificate authority to the file.
-        with open(ca_file, 'w') as fp:
-            fp.write(certificate_authority)
-        # Update the trusted CAs on this system.
+        # Create a path to install CAs on Debian systems.
+        ca_path = '/usr/local/share/ca-certificates/{0}.crt'.format(name)
+        log('Writing CA certificate to {0}'.format(ca_path))
+        _write_file(ca_path, certificate_authority)
+        # Update the trusted CAs on this system (a time expensive operation).
         check_call(['update-ca-certificates'])
-        message = 'Generated ca-certificates.crt for {0}'.format(name)
-        hookenv.log(message)
+        log('Generated ca-certificates.crt for {0}'.format(name))
+
+
+def _ensure_directory(path):
+    '''Ensure the parent directory exists creating directories if necessary.'''
+    directory = os.path.dirname(path)
+    if not os.path.isdir(directory):
```

Comment:
Alternatively you can also call os.makedirs(path, exist_ok=True) and you dont have to guard against it already existing. This implementation is fine, just food for thought.